File: src/main/java/org/luxons/sevenwonders/game/effects/Discount.java
package org.luxons.sevenwonders.game.effects;

import java.util.ArrayList;
import java.util.List;

import org.luxons.sevenwonders.game.RelativePlayerPosition;
import org.luxons.sevenwonders.game.boards.Board;
import org.luxons.sevenwonders.game.boards.TradingRules;
import org.luxons.sevenwonders.game.resources.ResourceType;

public class Discount extends InstantEffect {

    private final List<ResourceType> resourceTypes = new ArrayList<>();

    private final List<RelativePlayerPosition> targets = new ArrayList<>();

    private int discountedPrice = 1;

    public List<ResourceType> getResourceTypes() {
        return resourceTypes;
    }

    public List<RelativePlayerPosition> getTargets() {
        return targets;
    }

    public int getDiscountedPrice() {
        return discountedPrice;
    }

    public void setDiscountedPrice(int discountedPrice) {
        this.discountedPrice = discountedPrice;
    }

    @Override
    public void apply(Board board, Board leftNeighbourBoard, Board rightNeighbourBoard) {
        TradingRules rules = board.getTradingRules();
        for (ResourceType type : resourceTypes) {
            for (RelativePlayerPosition target : targets) {
                rules.setCost(type, target, discountedPrice);
            }
        }
    }
}

File: sw-engine/src/main/kotlin/org/luxons/sevenwonders/engine/effects/Discount.kt
package org.luxons.sevenwonders.engine.effects

import kotlinx.serialization.Serializable
import org.luxons.sevenwonders.engine.boards.Board
import org.luxons.sevenwonders.engine.data.serializers.ResourceTypesSerializer
import org.luxons.sevenwonders.model.resources.Provider
import org.luxons.sevenwonders.model.resources.ResourceType

@Serializable
internal data class Discount(
    @Serializable(with = ResourceTypesSerializer::class)
    val resourceTypes: List<ResourceType> = emptyList(),
    val providers: List<Provider> = emptyList(),
    val discountedPrice: Int = 1,
) : InstantOwnBoardEffect() {

    public override fun applyTo(board: Board) {
        val rules = board.tradingRules
        for (type in resourceTypes) {
            providers.forEach { rules.setCost(type, it, discountedPrice) }
        }
    }
}
